File: python/Letter_Combinations_of_a_Phone_Number.py
class Solution:
    # @return a list of strings, [s1, s2]
    def letterCombinations(self, digits):
        d = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        cands = [""]
        for digit in digits:
            if digit == None or ord(digit) < ord('2') or ord(digit) > ord('9'):
                continue
            tmp = []
            chars = d[ord(digit)-ord('2')]
            for cand in cands:
                for c in chars:
                    tmp.append(cand+c)
            cands = tmp
        return cands

File: cpp/Letter_Combinations_of_a_Phone_Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        string combs[8] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> ret;
        ret.push_back("");
        vector<string> tmp;
        int len = digits.length();
        if (len == 0) return ret;
        for (int n = 0; n < len; n ++) {
            string comb = combs[digits[n]-50]; 
            for (int i = 0; i < ret.size(); i ++){
                for (int j = 0; j < comb.length(); j ++){
                    string cand(ret[i]);
                    cand.append(1, comb[j]);
                    tmp.push_back(cand);
                }
            }
            ret = vector<string>(tmp);
            tmp.clear();
        }
        return ret;
    }
};
